### PR TITLE
Share instance of ClassTag in a ClassValue based cache

### PR DESCRIPTION
**Pull request by retronym**
A dynamic alternative to #7862

### CODE REVIEW

**Comment by lrytz** [comment]
Also a simpler alternative :-) Looks good.

**Comment by retronym** [comment]
I'll take a look at the compiler performance (and profiles thereof) [of the second commit](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/2489/console) to see if the cost of `ClassValue` hashmap lookup is perceptible or not.

**Comment by retronym** [comment]
The profile recorded one sample (of 5966) with `ClassValue.get` on the stack. The recorded times were a tiny bit slower than the previous commit, but there is some fluctuation due to JIT variance in those numbers.

We could create a way to enable the old behaviour with a system property to be more conservative.

I should also do a direct microbenchmark to see how the worst case looks.

**Comment by retronym** [comment]
The new benchmark actually improves (slightly) with the `ClassValue` cache.

```
[info] Benchmark                         Mode  Cnt   Score   Error  Units
[info] ClassTagBenchmark.lookupClassTag  avgt   20  53.383 ± 0.638  ns/op // Baseline
[info] ClassTagBenchmark.lookupClassTag  avgt   20  37.261 ± 0.308  ns/op // Optimize ClassTag.apply
...
[info] ClassTagBenchmark.lookupClassTag  avgt   20  35.012 ± 0.216  ns/op // Share instance of ClassTag ...
```

File: src/library/scala/reflect/ClassTag.scala [hunk 1]
```diff
@@ -111,29 +111,36 @@ object ClassTag {
...
-      case java.lang.Double.TYPE    => ClassTag.Double.asInstanceOf[ClassTag[T]]
-      case java.lang.Boolean.TYPE   => ClassTag.Boolean.asInstanceOf[ClassTag[T]]
-      case java.lang.Void.TYPE      => ClassTag.Unit.asInstanceOf[ClassTag[T]]
-      case ObjectTYPE               => ClassTag.Object.asInstanceOf[ClassTag[T]]
-      case NothingTYPE              => ClassTag.Nothing.asInstanceOf[ClassTag[T]]
-      case NullTYPE                 => ClassTag.Null.asInstanceOf[ClassTag[T]]
-      case _                        => new GenericClassTag[T](runtimeClass1)
-    }
+  def apply[T](runtimeClass1: jClass[_]): ClassTag[T] = runtimeClass1 match {
+    case x if x.isPrimitive =>  primitiveClassTag(runtimeClass1)
```
Comment (mkeskells) [inline src/library/scala/reflect/ClassTag.scala]: is it better to have this logic in the `computeValue`
depends on the frequency of lookup etc

Ideally we could add the special values to the cache on its construction, but I can see a hook to allow that in `ClassValue`

Comment (retronym) [reply src/library/scala/reflect/ClassTag.scala]: That sounds more regular, I've added a commit.

It slightly improves the lookup speed of refererence classtags.

```
[info] ClassTagBenchmark.lookupClassTag  avgt   20  33.653 ± 0.644  ns/op
```

**Comment by retronym** [comment]
I'm also applying this technique to `TypeTag`-s: https://github.com/scala/scala/pull/8112

**Comment by lrytz** [comment]
That seems ready. @retronym, squash and merge?

**Comment by lrytz** [comment]
And: it was originally against 2.12.x, then you moved it to 2.13, do you remember why?

**Comment by lrytz** [comment]
Hmm, could this prevent classes from being garbage collected?

**Comment by lrytz** [comment]
Ah, `ClassValue` uses weak references.

**Comment by retronym** [comment]
Squashed and retargeted to 2.12.x

**Comment by retronym** [comment]
"validate-main — No corresponding job found on Jenkins. Failed to launch? Try /rebuild " looks like a bug in scabot when retargetting PRs from 2.13.x to 2.13.x.

**Comment by mkurz** [comment]
@retronym @lrytz This never made it into `2.13.x` even though in #8242 it says `PICK`.
Will you forward port this finally or isn't it worth it? If you do, you also should forward port #9276 and #9361 afterwards to fix a classloader leak this patch here introduced.

**Comment by retronym** [comment]
Thanks for the reminder. I'll forward port it this week with the subsequent patch to avoid the leak.

**Comment by retronym** [comment]
Forward port proposed in #9487

**Pull request merged**